File: exemplos_aula09/java/ex4_prod_cons.java
import java.util.concurrent.*;

public class ex4_prod_cons {
	static int num_threads = 20;
	static int producers = 3;
	static int consumers = 3;
	static volatile int in = 0;
	static volatile int out = 0;
	static volatile int buffer[] = new int[num_threads];

	static class producer implements Runnable {
		Semaphore mutex;
		Semaphore empty;
		Semaphore full;
		private int tid;
		private int item = 0;

		public producer(int id, Semaphore mtx, Semaphore emp, Semaphore ful){
			this.tid = id;
			mutex = mtx;
			empty = emp;
			full = ful;
		}

		public void run(){
			while (true) {
				item++;
				try {
					empty.acquire();
					mutex.acquire();
					buffer[in] = item;
					System.out.println("produtor " + tid + " escrevendo na pos " + in);
					in = (in + 1) % num_threads;
					mutex.release();
					full.release();
				}catch (InterruptedException e) {
				}
			}
		}
	}

	static class consumer implements Runnable {
		Semaphore mutex;
		Semaphore empty;
		Semaphore full;
		private int tid;
		private int item = 0;

		public consumer(int id, Semaphore mtx, Semaphore emp, Semaphore ful){
			this.tid = id;
			mutex = mtx;
			empty = emp;
			full = ful;
		}

		public void run(){
			while (true) {
				item++;
				try {
					full.acquire();
					mutex.acquire();
					item = buffer[in];
					System.out.println("consumidor " + tid + " retirando da pos " + out);
					out = (out + 1) % num_threads;
					mutex.release();
					empty.release();
				}catch (InterruptedException e) {
				}
			}
		}
	}

	public static void main(String[] args) {
		Semaphore mutex = new Semaphore(1);
		Semaphore empty = new Semaphore(num_threads);
		Semaphore full = new Semaphore(0);
		int i;

		for (i = 0; i < producers; i++){
			Thread t = new Thread(new producer(i, mutex, empty, full));
			t.start();
		}
		for (i = 0; i < consumers; i++){
			Thread t = new Thread(new consumer(i, mutex, empty, full));
			t.start();
		}
	}
}

File: exemplos_aula09/c/ex4_prod_cons.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>

#define N		20
#define PRODUCERS	3
#define CONSUMERS	3

sem_t mutex, empty, full;
int in=0, out=0, buffer[N];

void *producer(void *arg){
	int item, tid;

	tid = (int)(long int)arg;
	while(1){
		item=random();
		sem_wait(&empty);			/* wait() ou p() ou down() */
		sem_wait(&mutex);			/* wait() ou p() ou down() ou lock() */
		buffer[in]=item;
		printf("produtor %d escrevendo na pos %d\n", tid, in);
		in=(in+1)%N;
		sem_post(&mutex);			/* post() ou v() ou up() ou signal() ou unlock() */
		sem_post(&full);			/* post() ou v() ou up() ou signal() */
	}
}

void *consumer(void *arg){
	int item, tid;

	tid = (int)(long int)arg;
	while(1){
		sem_wait(&full);			/* wait() ou p() ou down() */
		sem_wait(&mutex);			/* wait() ou p() ou down() */
		item=buffer[out];
		printf("consumidor %d retirando da pos %d\n", tid, out);
		out=(out+1)%N;
		sem_post(&mutex);			/* post() ou v() ou up() ou signal() */
		sem_post(&empty);			/* post() ou v() ou up() ou signal() */
		item=0;
	}
}

int main(void){
	long int i;
	pthread_t producers[PRODUCERS], consumers[CONSUMERS];
	
	sem_init(&mutex,0,1);
	sem_init(&empty,0,N);
	sem_init(&full,0,0);

	for(i = 0; i < PRODUCERS; i++)
		pthread_create(&producers[i], NULL, producer, (void *)i);
	for(i = 0; i < CONSUMERS; i++)
		pthread_create(&consumers[i], NULL, consumer, (void *)i);

	pthread_exit(NULL);	
	return(0);
}

